### ISSUE TITLE
executing shell command from a notebook cell

### ISSUE THREAD

**Issue by udibr** [facebook/iTorch#10]
In ipython notebook you can:

starting a cell with
%%bash 

or starting a single shell line with "!"

**Comment by ajtulloch** [comment]
Lua has `os.execute` if this is blocking you?

**Comment by udibr** [comment]
yes, my usage of notebooks is to combine documentation with code and shell commands.
A single run through all the cells of the notebook can perform all the operations needed.

In other otherwords, I expect shell command to be as easy to use as running code (torch/python).
Also in general cell magic (that start with %% or line magic that start with %) have many powerful options that I usually use in ipython notebook.

In addition, with ipython you can combine shell single line commands inside python command which is even better. For example a python list of all files in current directory can be done as follows:

   var = !ls -1 

I guess that the better solution is to combine all languages (torch, python and shell) together in single meta-engine and this is what jupyter is working on but I dont see torch in their future plans

I understand that starting a line with "!" or "%%" may cause problems with torch syntax but then just allow usage of a different escape varaible.

**Comment by soumith** [comment]
we could include the shell semantics that are in trepl.
in trepl, prefixing $ will execute shell.
> $ ls

https://github.com/torch/trepl/blob/master/init.lua#L531

**Comment by Atcold** [comment]
I agree with @udibr. The power of Jupyter/IPython is the integration of the shell commands with the host language, where you can use also

```
%mkdir $folder
```
for making a new directory with the name contained in the variable `folder`, and you can return the output of the shell back in your program (like `files = !ls -1` mentioned above).

Moreover, the `%bookmark` magic function is pretty awesome, letting you to jump around the filesystem very easily.
So, it would be great having these integrated, somehow. Even though, I don't have a clue if this is even possible.

**Comment by facebook-github-bot-0** [comment]
Thank you for reporting this issue and appreciate your patience. We've  notified the core team for an update on this issue. We're looking for a response within the next 30 days or the issue may be closed.

**Comment by ajdroid** [comment]
Any movement on this?

**Comment by wang3t** [comment]
Please add magic like that in ipython (https://ipython.org/ipython-doc/3/interactive/magics.html)

in iTorch, 
  $ls => unexpected symbol near '$'
  %%time => unexpected symbol near '%'
in trepl, 
  $ls  => it executes but output a block of black output (font and background are all in black)
